### PR TITLE
TemporalDropout1D

### PR DESCRIPTION
**Pull request by halidziya**
Removing words and estimating from context is one of the recent language modeling tasks. This kind of noise could be useful as regularization as well.

### CODE REVIEW

**Comment by halidziya** [comment]
@googlebot I signed it!

File: tensorflow/python/keras/layers/core.py [hunk 1]
```diff
@@ -236,6 +236,39 @@ def _get_noise_shape(self, inputs):
...
+    Same as input.
+  """
+
+  def __init__(self, rate, **kwargs):
+    super(TemporalDropout1D, self).__init__(rate, **kwargs)
+    self.input_spec = InputSpec(ndim=3)
+
+  def _get_noise_shape(self, inputs):
+    input_shape = array_ops.shape(inputs)
+    noise_shape = (input_shape[0], input_shape[2], 1)
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/core.py:270]: What's the diff between `TemporalDropout1D` and `SpatialDropout1D`? By looking at the implementation it looks like tf.keras.layers.SpatialDropout1D is actually temporal (i.e., applies to time_steps)

**Comment by gbaned** [comment]
@halidziya Can you please check reviewer comments and keep us posted. Thanks!

File: tensorflow/python/keras/layers/core.py (hunk 1 above)
Comment (halidziya) [reply tensorflow/python/keras/layers/core.py:270]: Hi, 
    There is a transpose difference. 
Here this is an example code which shows  SpatialDropout1D is droping the features not time points

SpatialDropout1D(0.5)(np.ones((10,10,10))*2.0, training=True)[0, 0]

[0., 4., 4., 0., 4., 4., 4., 4., 4., 4.]

Comment (tanzhenyu) [reply tensorflow/python/keras/layers/core.py:270]: The `time_step` axis is axis=1, so,
```python
SpatialDropout1D(0.5)(np.ones((10,10,10))*2.0, training=True)[0, 0]
<tf.Tensor: shape=(10, 10), dtype=float32, numpy=
array([[0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.],
       [0., 4., 0., 0., 4., 4., 4., 0., 4., 0.]], dtype=float32)>
```
means the dropout is applied along the `time_step` axis, i.e., feature 0 is dropped out, so output[0][:][0] are all 0, feature 1 is not dropped out, so output[0][:][1] are all 4., correct?

Comment (halidziya) [reply tensorflow/python/keras/layers/core.py:270]: If time_step axis is 1 , yes it drops one feature for all time steps. But for temporal dropout, we would like to drop all feature for a given time step. This is similar to removing a word from text.

Comment (tanzhenyu) [reply tensorflow/python/keras/layers/core.py:270]: That's not TemporalDropout then. This is a useful feature though, which I think you can simply achieve it by calling `tf.keras.layers.Dropout(0.5, noise_shape=[None, 10, 1])(inp, training=True)`

Without any subclassing or having a new endpoint.

**Comment by tanzhenyu** [comment]
Closing this PR now. See discussions above. Let us know if you have other concerns.
